### PR TITLE
Add Mopeka BLE and Mopeka Pro Check BLE Sensor

### LINKED ISSUE #1445: Support Mopeka Pro Check LP gas tank BLE sensor
**Describe the problem you have/What new integration you would like**

Add support for Mopeka Pro Check LP gas tank sensor.  This is a battery powered BLE sensor that can be mounted on a propane tank and will report the tank level.   sensor here: https://mopeka.com/product/mopeka-check-pro-lp-sensor/

Sensor can report temperature, battery level, ultra-sonic measured distance, and that can be converted into a percentage for a given tank.   Sensor also has information regarding quality of measurement (0-3 rating) with 3 being highest quality.  It also has a sync button that is used for enroll in their app.  This could be used by the integration to help the user identify the MAC address of the device so that they can create a fully defined sensor.  

**Please describe your use case for this integration and alternatives you've tried:**

I have created a native HA (HACS) integration but it requires that your HA have BLE and that your HA instance be within range of the sensor.  This EspHome based solution would allow for more remote usage scenarios over wifi.  
HA Sensor <https://github.com/spbrogan/sensor.mopeka_pro_check>
Python BLE component <https://github.com/spbrogan/mopeka_pro_check>

**Additional context**

### PR DESCRIPTION
**Pull request by spbrogan**
# What does this implement/fix? 

Add support for Mopeka Pro Check LP BLE sensor and adds generic component that helps find the BLE MAC address.  
This is a similar pattern to other BLE advertisement based sensors.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <https://github.com/esphome/feature-requests/issues/1445>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1564

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml
sensor:
  # Example using 20lb vertical propane tank.
  - platform: mopeka_pro_check
    mac_address: D3:75:F2:DC:16:91
    tank_type: 20LB_V
    temperature:
      name: "Propane test temp"
    level:
      name: "Propane test level"
    distance:
      name: "Propane test distance"
    battery_level:
      name: "Propane test battery level"

  # Custom example - user defined empty / full points
  - platform: mopeka_pro_check
    mac_address: D3:75:F2:DC:16:91
    tank_type: CUSTOM
    custom_distance_full: 40cm
    custom_distance_empty: 10mm
    temperature:
      name: "Propane test temp"
    level:
      name: "Propane test level"
    distance:
      name: "Propane test distance"
    battery_level:
      name: "Propane test battery level"

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by spbrogan** [comment]
I have done minor refactor since it was last functionally tested.  I will run tests to make sure both components work before checking that task off.

**Comment by spbrogan** [comment]
> I have done minor refactor since it was last functionally tested. I will run tests to make sure both components work before checking that task off.

this works for me locally.  I think the PR is ready to go.  Please provide any feedback.

**Comment by spbrogan** [comment]
do you want me to resolve the conflict as it is just the test code which should be trivial?  I assume when you are ready to merge you would resolve the conflicts (as more could come up before you choose to merge this).

**Comment by spbrogan** [comment]
What needs to happen to get this merged?  Is there a step in the process I am missing?

**Comment by sdellysse** [comment]
echoing @spbrogan's comment: I'm building a system with a few ESPHome devices, and having this in-main would be really nice

**Comment by spbrogan** [comment]
I just force pushed to the PR.  Only 2 changes.  
1. rebased on latest dev to remove old merge commits
2. updated the defined values for min/max for standard US propane tanks so that the percent calculation more closely aligns with the sensor supplied app.

**Comment by spbrogan** [comment]
can this get merged.  I just got notified my documentation PR will be closed if no activity.  Is there something more that is needed?

**Comment by spbrogan** [comment]
This has been done since Oct.  Tested, documented, and even requested by other community members.  I can't figure out what needs to happen to get this merged or how to draw the necessary attention to it.  I have noticed plenty of similar new integrations have been merged much faster so again I am looking for some feedback for how this project prioritizes contributions and pull requests.  Should I ask on discord? Do the "maintainers" not have BBQs and thus don't track their propane levels?  :)

**Comment by helenclarko** [comment]
Who needs to see this to get it merged in?

File: esphome/components/mopeka_pro_check/sensor.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    except cv.Invalid:
+        pass
+    try:
+        cm = _distance_cm(value)
+        return cm * 10
+    except cv.Invalid:
+        pass
+
+    meters = cv.distance(value)
+    return meters * 100
```
Comment (jesserockz) [inline esphome/components/mopeka_pro_check/sensor.py]: `cv.distance` takes care of all conversions. You just need to multiply the value by 1000 to get mm at the end.

```suggestion
def small_distance(value):
    """small_distance is stored in mm"""
    meters = cv.distance(value)
    return meters * 1000
```

Comment (spbrogan) [reply esphome/components/mopeka_pro_check/sensor.py]: My understanding: cv.distance() only supports meters.  that means a user must enter the values in meters which is awkward.  
https://github.com/esphome/esphome/blob/0ec84be5dacb1a9d0c3cc5c5da5de9a3e9b14df9/esphome/config_validation.py#L848

So the small_distance function allows a user to enter in mm, cm, or meters. 

Am i reading it wrong?

Comment (spbrogan) [reply esphome/components/mopeka_pro_check/sensor.py]: regardless...there is a bug in that line 48 is only doing meters * 100.  So if a user had entered the distance in meters they would not get the right values.  I'll fix that once we have alignment on the above.  Thanks

Comment (jesserockz) [reply esphome/components/mopeka_pro_check/sensor.py]: > Am i reading it wrong?

Yes :smile: 

It means that the base unit and returned value is meters. A user can still put `500mm`, and then the validator will return `0.5` as the actual value (which is why you need the `* 1000`

Comment (jesserockz) [reply esphome/components/mopeka_pro_check/sensor.py]: `float_with_unit` handles prefixes and multiplications on the `m` (for meters)

File: esphome/components/mopeka_pro_check/sensor.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+CONF_TANK_TYPE = "tank_type"
+CONF_CUSTOM_DISTANCE_FULL = "custom_distance_full"
+CONF_CUSTOM_DISTANCE_EMPTY = "custom_distance_empty"
+
+ICON_PROPANE_TANK = "mdi:propane-tank"
+
+TANK_TYPE_CUSTOM = "CUSTOM"
+
+UNIT_MILLIMETER = "mm"
+
```
Comment (jesserockz) [inline esphome/components/mopeka_pro_check/sensor.py:29]:
```suggestion


```

**Comment by helenclarko** [comment]
Good to go now then?

**Review by jesserockz — APPROVED** [review]

File: esphome/components/mopeka_pro_check/sensor.py [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+CONFIG_SCHEMA = (
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(MopekaProCheck),
+            cv.Required(CONF_MAC_ADDRESS): cv.mac_address,
+            cv.Optional(CONF_CUSTOM_DISTANCE_FULL): small_distance,
+            cv.Optional(CONF_CUSTOM_DISTANCE_EMPTY): small_distance,
+            cv.Optional(CONF_TANK_TYPE, default="20LB_V"): cv.enum(
+                CONF_SUPPORTED_TANKS_MAP, upper=True
+            ),
```
Comment (jesserockz) [inline esphome/components/mopeka_pro_check/sensor.py]: I actually noticed this stand out more when reading the docs PR, should there even be a default here? I think should be required for everyone.
```suggestion
            cv.Required(CONF_TANK_TYPE): cv.enum(
                CONF_SUPPORTED_TANKS_MAP, upper=True
            ),
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Comment by helenclarko** [comment]
What does this need now?

**Review by jesserockz — APPROVED** [review]

**Pull request merged**